lib/provider/settings_provider.dart

```
@@ -252,11 +252,11 @@ class SettingsProvider with ChangeNotifier {
   set lastPushToken(String? value) =>
       preferences!.setString(KEY_LAST_PUSH_TOKEN, value!);*/
 
-  //Token
+  // Token. If token is invalid, return null.
   JWToken? get fduholeToken {
     if (preferences!.containsKey(KEY_FDUHOLE_TOKEN)) {
       try {
-        return JWToken.fromJson(
+        return JWToken.fromJsonWithVerification(
```

If we need verification, why not just make `access` and `refresh` non-nullable. Do we need to write our own methods?

---

It is a practice to throw our own error type `BadTokenException` rather than a `NullException`, so we can catch it during logging in or signing up, and distinguish it from other null exception.